File: Python3/538.Convert_BST_to_Greater_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.total = 0
    def convertBST(self, root: TreeNode) -> TreeNode:
        if root != None:
            self.convertBST(root.right)
            self.total += root.val
            root.val = self.total
            self.convertBST(root.left)
        return root

#recursive but no global variable
class Solution:
    def convertBST(self, root: TreeNode) -> TreeNode:
        self.convertBSTHelper(root, 0)
        return root
    
    def convertBSTHelper(self, root: TreeNode, total: int) -> int:
        if root == None: 
            return total 

        right = self.convertBSTHelper(root.right, total)        
        root.val += right
        left = self.convertBSTHelper(root.left, root.val)
        return left

#unfinished iterative
class Solution:
    def convertBST(self, root: TreeNode):
        if root == None:
            return root
        
        stack = []
        stack.append(root)
        total = 0

        while len(stack) != 0:
            node = stack.pop()
            if node == None:
                continue
            
            stack.append(node.right)

            node.val += total
            total = node.val
            stack.append(node.left)
            

        return root

File: Java/538.Convert_BST_to_Greater_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int total = 0;

    public TreeNode convertBST(TreeNode root) {
        if (root != null) {
            convertBST(root.right);
            total += root.val;
            root.val = total;
            convertBST(root.left);
        }
        
        return root;
    }
}

class Solution {
    public TreeNode convertBST(TreeNode root) {
        if (root == null) return root;
        convertBSTHelper(root, 0);
        return root;
    }
    
    private int convertBSTHelper(TreeNode root, int total) {
        if (root == null) return total;

        int right = convertBSTHelper(root.right, total);
        root.val += right;
        int left = convertBSTHelper(root.left, root.val);
        
        return left;
    }
}

